src/lib/workouts/pushup.ts

```diff
@@ -4,40 +4,104 @@ import Workout from '.';
 export default class Pushup extends Workout {
 	private isUp: boolean;
 	private pushups: number;
+	private average_left_wrist: { x: number; y: number } | null;
+	private average_right_wrist: { x: number; y: number } | null;
+	private iterations: number;
 	constructor() {
 		super();
 		this.isUp = true;
+		this.iterations = 0;
 		this.pushups = 0;
+		this.average_left_wrist = null;
+		this.average_right_wrist = null;
 	}
 	public onFrame(poses: Pose[]): string {
+		this.iterations++;
 		const keypoints = poses[0].keypoints;
-		const left_shoulder = keypoints.find((keypoint) => keypoint.name === 'left_shoulder');
-		const right_shoulder = keypoints.find((keypoint) => keypoint.name === 'right_shoulder');
-		const left_elbow = keypoints.find((keypoint) => keypoint.name === 'left_elbow');
-		const right_elbow = keypoints.find((keypoint) => keypoint.name === 'right_elbow');
-		const left_hip = keypoints.find((keypoint) => keypoint.name === 'left_hip');
-		const right_hip = keypoints.find((keypoint) => keypoint.name === 'right_hip');
+		const left_shoulder = keypoints.find((keypoint) => keypoint.name === 'left_shoulder')!;
+		const right_shoulder = keypoints.find((keypoint) => keypoint.name === 'right_shoulder')!;
+		const left_elbow = keypoints.find((keypoint) => keypoint.name === 'left_elbow')!;
+		const right_elbow = keypoints.find((keypoint) => keypoint.name === 'right_elbow')!;
+		const left_wrist = keypoints.find((keypoint) => keypoint.name === 'left_wrist')!;
+		const right_wrist = keypoints.find((keypoint) => keypoint.name === 'right_wrist')!;
 
-		const left_shoulder_to_elbow = Math.round(
-			Math.sqrt((left_shoulder.x - left_elbow.x) ** 2 + (left_shoulder.y - left_elbow.y) ** 2)
+		// keep a running average
+		if (this.average_left_wrist === null) {
+			this.average_left_wrist = left_wrist;
+		} else {
+			const distance_from_average = Math.sqrt(
+				(left_wrist.x - this.average_left_wrist.x) ** 2 +
+					(left_wrist.y - this.average_left_wrist.y) ** 2
+			);
+			this.average_left_wrist.x += (left_wrist.x - this.average_left_wrist.x) / this.iterations;
+			this.average_left_wrist.y += (left_wrist.y - this.average_left_wrist.y) / this.iterations;
+			if (distance_from_average > 10) {
+				return "Frame rejected; don't move your wrist";
+			}
+		}
+		if (this.average_right_wrist === null) {
+			this.average_right_wrist = right_wrist;
+		} else {
+			const distance_from_average = Math.sqrt(
+				(left_wrist.x - this.average_left_wrist.x) ** 2 +
+					(left_wrist.y - this.average_left_wrist.y) ** 2
+			);
+			this.average_right_wrist.x += (right_wrist.x - this.average_right_wrist.x) / this.iterations;
+			this.average_right_wrist.y += (right_wrist.y - this.average_right_wrist.y) / this.iterations;
+			if (distance_from_average > 10) {
+				return "Frame rejected; don't move your wrist";
+			}
+		}
+		const nose = keypoints.find((keypoint) => keypoint.name === 'nose')!;
```

Comment:
For clarity move the keypoint up with the rest of the keypoints at the top please

Reply:
I'd rather have the code end early when the frame would be rejected either way